modules/tournament/src/main/TournamentScheduler.scala

```diff
@@ -349,35 +350,36 @@ Thank you all, you rock!"""
           }
         ).flatten
       }
-    ).flatten map { plan =>
-        plan.copy(schedule = Schedule addCondition plan.schedule)
-      }
+    ).flatten filter { _.schedule.at.isBefore(rightNow) }
   }
 
-  private[tournament] def pruneConflicts(scheds: List[Schedule], allPlans: List[Plan]): List[Plan] = {
-    allPlans.foldLeft(List[Plan]()) {
-      case (plans, p) if p.schedule.at.isBeforeNow => plans
-      case (plans, p) if overlaps(p.schedule, scheds) => plans
-      case (plans, p) if overlaps(p.schedule, plans.map(_.schedule)) => plans
-      case (plans, p) => p :: plans
+  private[tournament] def pruneConflicts(scheds: List[Tournament], allPlans: List[Plan]) = {
+    val tours = new ListBuffer[Tournament]
+    allPlans foreach { plan =>
+      val schedule = Schedule.addCondition(plan.schedule)
+      val minutes = Schedule.durationFor(schedule)
+      val t = plan.build.foldRight(Tournament.schedule(schedule, minutes)) { _(_) }
+      if (!overlaps(t, tours) && !overlaps(t, scheds)) tours += t
     }
+    tours.toList
```

Comment:
why the mutability here?

Reply:
Well the current code reverses the order of the tournaments during the filter, which I find consider an unwelcome side effect that made it harder for me to reason about the code. 

So that’s why I’m changing from a foldLeft with a List.

You could foldLeft with a ListBuffer but the instance itself doesn’t change so passing along the ref in fold is counterintuitive. ListBuffer is the natural choice if we want a List output. You can also instantiate with List.newBuilder.

Reply:
Yes but I mean, we had immutable, functional code that works and has no performance issue, and you replaced it with mutable code with side effects. `ListBuffer` is a mutable structure, pushing to it is a side effect, and the whole block becomes imperative. You know I can't let that pass.

Reply:
Alright, will adjust.